mongo/session.go

```diff
@@ -26,150 +26,62 @@ var ErrWrongClient = errors.New("session was not created by this client")
-// the provided callback. The other is to use NewSessionContext to explicitly create a SessionContext.
-type SessionContext interface {
-	context.Context
-	Session
-}
-
-type sessionContext struct {
-	context.Context
-	Session
+// Implementations of Session are not safe for concurrent use by multiple
+// goroutines.
+type Session struct {
+	clientSession       *session.Client
+	client              *Client
+	deployment          driver.Deployment
+	didCommitAfterStart bool // true if commit was called after start with no other operations
 }
 
-type sessionKey struct {
-}
+type sessionKey struct{}
 
-// NewSessionContext creates a new SessionContext associated with the given Context and Session parameters.
-func NewSessionContext(ctx context.Context, sess Session) SessionContext {
-	return &sessionContext{
-		Context: context.WithValue(ctx, sessionKey{}, sess),
-		Session: sess,
-	}
+// ContextWithSession creates a new SessionContext associated with the given
+// Context and Session parameters.
+func ContextWithSession(parent context.Context, sess *Session) context.Context {
+	return context.WithValue(parent, sessionKey{}, sess)
 }
 
 // SessionFromContext extracts the mongo.Session object stored in a Context. This can be used on a SessionContext that
 // was created implicitly through one of the callback-based session APIs or explicitly by calling NewSessionContext. If
 // there is no Session stored in the provided Context, nil is returned.
-func SessionFromContext(ctx context.Context) Session {
+func SessionFromContext(ctx context.Context) *Session {
 	val := ctx.Value(sessionKey{})
 	if val == nil {
 		return nil
 	}
 
-	sess, ok := val.(Session)
+	sess, ok := val.(*Session)
 	if !ok {
 		return nil
 	}
 
 	return sess
 }
 
-// Session is an interface that represents a MongoDB logical session. Sessions can be used to enable causal consistency
-// for a group of operations or to execute operations in an ACID transaction. A new Session can be created from a Client
-// instance. A Session created from a Client must only be used to execute operations using that Client or a Database or
-// Collection created from that Client. Custom implementations of this interface should not be used in production. For
-// more information about sessions, and their use cases, see
-// https://www.mongodb.com/docs/manual/reference/server-sessions/,
-// https://www.mongodb.com/docs/manual/core/read-isolation-consistency-recency/#causal-consistency, and
-// https://www.mongodb.com/docs/manual/core/transactions/.
-//
-// Implementations of Session are not safe for concurrent use by multiple goroutines.
-type Session interface {
-	// StartTransaction starts a new transaction, configured with the given options, on this
-	// session. This method returns an error if there is already a transaction in-progress for this
-	// session.
-	StartTransaction(...*options.TransactionOptions) error
```

Comment:
We should move all of the godoc comments from this interface to the `Session` methods.

Reply:
Good catch 👍